File: OpenBurnBarMobile/Views/Components/Pro/LockedFeatureVeil.swift
import SwiftUI
import OpenBurnBarCore

// MARK: - Locked Feature Veil
//
// Pro vocabulary — frosted mercury veil over a locked feature. The content
// behind is rendered blurred-but-visible (a teaser) so the user *sees*
// what they're missing — the I-WANT-THAT moment. Centered CTA inside an
// obsidian foil card.
//
// Two content modes:
//   • `.legacy` — the original headline/detail/MercuryCrest anatomy (still used
//     by the cloud-tier Insights / Streams teasers). Default `ctaLabel`
//     "Open Cloud".
//   • `.feature(GatedFeature)` — the tier+feature-aware anatomy: the required
//     tier's glowing holographic crest, the serif `publicName`, the
//     `oneLineBenefit`, the catalog `benefitBullets`, and a
//     "Unlock with <Tier>" FoilCTA, with an "Available on <Tier>" subheader.
//     Driven entirely by the honesty-checked catalog copy.
//
// Suppressed when the user is a member; the host view should just render
// `background` directly in that case.

struct LockedFeatureVeil<Background: View>: View {
    enum Content {
        case legacy(headline: String, detail: String, ctaLabel: String, icon: String)
        case feature(GatedFeature, priceLine: String?)
    }

    let content: Content
    let action: () -> Void
    let background: Background

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    // MARK: Designated init

    init(content: Content, action: @escaping () -> Void, background: Background) {
        self.content = content
        self.action = action
        self.background = background
    }

    // MARK: Legacy init (preserves existing call sites)

    init(
        headline: String,
        detail: String,
        ctaLabel: String = "Open Cloud",
        icon: String = "sparkle",
        action: @escaping () -> Void,
        @ViewBuilder background: () -> Background
    ) {
        self.content = .legacy(headline: headline, detail: detail, ctaLabel: ctaLabel, icon: icon)
        self.action = action
        self.background = background()
    }

    var body: some View {
        ZStack {
            // Blurred teaser content — visible enough to be evocative, dim
            // enough that the foil card reads as the focal point.
            background
                .blur(radius: 16)
                .saturation(0.78)
                .opacity(0.7)
                .allowsHitTesting(false)
                .accessibilityHidden(true)

            // Obsidian veil dims the teaser into the Pro palette.
            LinearGradient(
                colors: [
                    ProTheme.Palette.obsidian.opacity(0.55),
                    ProTheme.Palette.obsidian.opacity(0.82)
                ],
                startPoint: .top,
                endPoint: .bottom
            )
            .allowsHitTesting(false)

            if !reduceMotion {
                MercuryShimmerOverlay()
                    .blendMode(.plusLighter)
                    .opacity(0.4)
                    .allowsHitTesting(false)
            }

            switch content {
            case let .legacy(headline, detail, ctaLabel, icon):
                legacyBody(headline: headline, detail: detail, ctaLabel: ctaLabel, icon: icon)
            case let .feature(feature, priceLine):
                featureBody(feature: feature, priceLine: priceLine)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .accessibilityElement(children: .contain)
    }

    // MARK: Legacy anatomy

    @ViewBuilder
    private func legacyBody(headline: String, detail: String, ctaLabel: String, icon: String) -> some View {
        VStack(spacing: MobileTheme.Spacing.lg) {
            MercuryCrest(size: .large, shimmer: !reduceMotion)

            VStack(spacing: MobileTheme.Spacing.sm) {
                Text(headline)
                    .font(ProTheme.Typography.titleSerif)
                    .foregroundStyle(ProTheme.Palette.mercury)
                    .multilineTextAlignment(.center)
                    .fixedSize(horizontal: false, vertical: true)
                Text(detail)
                    .font(MobileTheme.Typography.body)
                    .foregroundStyle(ProTheme.Palette.mercury.opacity(0.72))
                    .multilineTextAlignment(.center)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .padding(.horizontal, MobileTheme.Spacing.xl)

            FoilCTAButton(title: ctaLabel, icon: icon, fillWidth: false, action: action)
                .frame(maxWidth: 340)
        }
        .padding(.horizontal, MobileTheme.Spacing.xl)
    }

    // MARK: Feature anatomy — the shared unlock content over the teaser

    @ViewBuilder
    private func featureBody(feature: GatedFeature, priceLine: String?) -> some View {
        ScrollView {
            FeatureUnlockContent(
                feature: feature,
                priceLine: priceLine,
                onUnlock: action,
                onDismiss: nil
            )
        }
        .scrollBounceBehavior(.basedOnSize)
    }
}

// MARK: - Preview

#Preview("Locked Feature Veil — legacy") {
    LockedFeatureVeil(
        headline: "Insights, surfaced.",
        detail: "Cross-agent patterns, weekly retros, and forecast cohorts — included with OpenBurnBar Cloud.",
        action: {}
    ) {
        VStack(spacing: 16) {
            ForEach(0..<4) { _ in
                RoundedRectangle(cornerRadius: 14)
                    .fill(MobileTheme.Colors.surface)
                    .frame(height: 80)
            }
        }
        .padding(20)
    }
}

#Preview("Locked Feature Veil — feature (Data Vault)") {
    LockedFeatureVeil(
        feature: GatedFeature.gatedFeature(.dataVault),
        priceLine: "From $24.99/mo",
        action: {}
    ) {
        VStack(spacing: 16) {
            ForEach(0..<4) { _ in
                RoundedRectangle(cornerRadius: 14)
                    .fill(MobileTheme.Colors.surface)
                    .frame(height: 80)
            }
        }
        .padding(20)
    }
}

File: android/app/src/main/java/com/openburnbar/ui/pro/LockedFeatureVeil.kt
@file:Suppress("MagicNumber")
// Compose layout literals (dp/sp/alpha); token-per-line extraction obscures UI structure.

package com.openburnbar.ui.pro

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.blur
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.openburnbar.ui.theme.LocalAuroraReduceMotion

/**
 * Pro vocabulary — frosted mercury veil over a locked feature. Content behind is
 * rendered blurred-but-visible so the user *sees* what they're missing.
 *
 * Two ways to drive it:
 *  • [LockedFeatureVeil] taking a [GatedFeature] — the tier+feature-aware
 *    unlock experience (spec §4.4): the required tier's holographic crest, the
 *    feature name, its one-liner, the "what you'll unlock" bullets, a single
 *    "Unlock with <Tier>" FoilCTAButton, and a quiet "Maybe later".
 *  • The legacy headline/detail overload — kept so existing call sites compile;
 *    new gates should pass a [GatedFeature].
 *
 * Skip rendering this and just show `background` when the user's tier satisfies
 * the feature's requirement.
 */
@Composable
fun LockedFeatureVeil(
    feature: GatedFeature,
    onUnlock: () -> Unit,
    modifier: Modifier = Modifier,
    livePrice: String? = null,
    onMaybeLater: (() -> Unit)? = null,
    background: @Composable () -> Unit,
) {
    Box(
        modifier = modifier.fillMaxSize(),
        contentAlignment = Alignment.Center,
    ) {
        BlurredTeaser(background = background)
        Box(
            modifier =
            Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState()),
            contentAlignment = Alignment.Center,
        ) {
            FeatureUnlockAnatomy(
                feature = feature,
                livePrice = livePrice,
                onUnlock = onUnlock,
                onMaybeLater = onMaybeLater,
                modifier = Modifier.padding(horizontal = 28.dp, vertical = 36.dp),
            )
        }
    }
}

/**
 * Legacy string-driven veil. Retained for entry points that have not yet been
 * migrated to a [GatedFeature]. Renders the MercuryCrest (no tier crest) and a
 * generic CTA. Prefer the [GatedFeature] overload for new work.
 */
@Composable
fun LockedFeatureVeil(
    headline: String,
    detail: String,
    onCta: () -> Unit,
    modifier: Modifier = Modifier,
    ctaLabel: String = "Open Cloud",
    background: @Composable () -> Unit,
) {
    val reduceMotion = LocalAuroraReduceMotion.current
    Box(
        modifier = modifier.fillMaxSize(),
        contentAlignment = Alignment.Center,
    ) {
        BlurredTeaser(background = background)
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(18.dp),
            modifier = Modifier.padding(horizontal = 28.dp),
        ) {
            MercuryCrest(size = MercuryCrestSize.Large, shimmer = !reduceMotion)
            Text(
                text = headline,
                style = ProTypography.titleSerif,
                color = ProPalette.mercury,
                textAlign = TextAlign.Center,
            )
            Text(
                text = detail,
                style = ProTypography.headlineSerif.copy(fontWeight = FontWeight.Normal),
                color = ProPalette.mercury.copy(alpha = 0.72f),
                textAlign = TextAlign.Center,
            )
            FoilCTAButton(
                title = ctaLabel,
                onClick = onCta,
                fillWidth = false,
                modifier = Modifier.widthIn(min = 220.dp, max = 320.dp),
            )
        }
    }
}

/** The blurred teaser of the real feature + the obsidian veil over it. */
@Composable
private fun BlurredTeaser(background: @Composable () -> Unit) {
    Box(
        modifier =
        Modifier
            .fillMaxSize()
            .blur(radius = 16.dp)
            .alpha(0.7f),
    ) {
        background()
    }
    Box(
        modifier =
        Modifier
            .fillMaxSize()
            .background(
                brush =
                Brush.verticalGradient(
                    colors =
                    listOf(
                        ProPalette.obsidian.copy(alpha = 0.55f),
                        ProPalette.obsidian.copy(alpha = 0.82f),
                    ),
                ),
            ),
    )
}
